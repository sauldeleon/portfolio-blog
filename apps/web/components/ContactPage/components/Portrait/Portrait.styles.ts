import Image from 'next/image'
import { css, keyframes, styled } from 'styled-components'

export const StyledPortraitContainer = styled.div`
  position: relative;
  width: 70px;
  height: 70px;

  ${({ theme }) => theme.media.up.md} {
    width: 140px;
    height: 140px;
  }
`

export const portraitStyles = css`
  overflow: hidden;
  border-radius: 50%;
  width: 70px;
  height: 70px;

  ${({ theme }) => theme.media.up.md} {
    width: 140px;
    height: 140px;
  }
`

const generateAnimation = (index: number, total: number) =>
  keyframes`
  ${css`
    ${(index / total) * 100}% {
      opacity: 0;
    }
    ${((index + 1) / total) * 100}% {
      opacity: 1;
    }
  `}
     
  `

export const StyledPortrait = styled(Image)<{
  $index: number
  $totalImages: number
}>`
  ${portraitStyles};
  position: absolute;
  opacity: ${({ $index }) => ($index === 0 ? 1 : 0)};

  ${({ $index, $totalImages }) =>
    $index > 0 &&
    css`
      animation: ${generateAnimation($index, $totalImages + 1)}
        ${($totalImages - 1) * 5}s infinite;
    `}
`
