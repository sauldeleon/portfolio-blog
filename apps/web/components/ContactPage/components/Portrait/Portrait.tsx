import { ImageProps } from 'next/image'

import { useClientTranslation } from '@web/i18n/client'

import { StyledPortrait } from './Portrait.styles'

interface PortraitProps
  extends Omit<ImageProps, 'src' | 'alt' | 'width' | 'height'> {
  onClick: () => void
}
export function Portrait({ onClick }: PortraitProps) {
  const { t } = useClientTranslation('contactPage')
  return (
    <StyledPortrait
      src="/assets/portrait-1.png"
      alt={t('profilePicture')}
      width={140}
      height={140}
      onClick={onClick}
    />
  )
}
