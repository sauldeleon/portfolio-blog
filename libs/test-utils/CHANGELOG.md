# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
