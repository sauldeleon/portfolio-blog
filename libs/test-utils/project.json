{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
