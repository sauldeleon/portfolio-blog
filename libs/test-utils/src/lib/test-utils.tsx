import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
