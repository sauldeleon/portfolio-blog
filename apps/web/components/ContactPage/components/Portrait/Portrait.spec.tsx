import { screen } from '@testing-library/react'

import { renderApp } from '@sdlgr/test-utils'

import { Portrait } from './Portrait'

describe('Portrait', () => {
  it('should render successfully', async () => {
    const { baseElement } = renderApp(<Portrait onClick={() => void 0} />)
    await screen.findAllByAltText('My profile picture')
    expect(baseElement).toMatchSnapshot()
  })
})
