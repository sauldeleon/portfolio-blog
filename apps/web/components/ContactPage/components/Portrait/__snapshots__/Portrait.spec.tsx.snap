// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portrait should render successfully 1`] = `
.c0 {
  position: relative;
  width: 70px;
  height: 70px;
}

.c1 {
  overflow: hidden;
  border-radius: 50%;
  width: 70px;
  height: 70px;
  position: absolute;
  opacity: 1;
}

.c2 {
  overflow: hidden;
  border-radius: 50%;
  width: 70px;
  height: 70px;
  position: absolute;
  opacity: 0;
  animation: cgzFdQ 25s infinite;
}

.c3 {
  overflow: hidden;
  border-radius: 50%;
  width: 70px;
  height: 70px;
  position: absolute;
  opacity: 0;
  animation: jGA-DsB 25s infinite;
}

.c4 {
  overflow: hidden;
  border-radius: 50%;
  width: 70px;
  height: 70px;
  position: absolute;
  opacity: 0;
  animation: hisaOC 25s infinite;
}

.c5 {
  overflow: hidden;
  border-radius: 50%;
  width: 70px;
  height: 70px;
  position: absolute;
  opacity: 0;
  animation: cLUxYQ 25s infinite;
}

.c6 {
  overflow: hidden;
  border-radius: 50%;
  width: 70px;
  height: 70px;
  position: absolute;
  opacity: 0;
  animation: rmUOo 25s infinite;
}

@media (min-width: 768px) {
  .c0 {
    width: 140px;
    height: 140px;
  }
}

@media (min-width: 768px) {
  .c1 {
    width: 140px;
    height: 140px;
  }
}

@media (min-width: 768px) {
  .c2 {
    width: 140px;
    height: 140px;
  }
}

@media (min-width: 768px) {
  .c3 {
    width: 140px;
    height: 140px;
  }
}

@media (min-width: 768px) {
  .c4 {
    width: 140px;
    height: 140px;
  }
}

@media (min-width: 768px) {
  .c5 {
    width: 140px;
    height: 140px;
  }
}

@media (min-width: 768px) {
  .c6 {
    width: 140px;
    height: 140px;
  }
}

<body>
  <div>
    <div
      class="c0"
    >
      <img
        alt="My profile picture"
        class="c1"
        data-nimg="1"
        decoding="async"
        height="140"
        loading="lazy"
        src="/_next/image?url=%2Fassets%2Fportrait-1.png&w=384&q=75"
        srcset="/_next/image?url=%2Fassets%2Fportrait-1.png&w=256&q=75 1x, /_next/image?url=%2Fassets%2Fportrait-1.png&w=384&q=75 2x"
        style="color: transparent;"
        width="140"
      />
      <img
        alt="My profile picture"
        class="c2"
        data-nimg="1"
        decoding="async"
        height="140"
        loading="lazy"
        src="/_next/image?url=%2Fassets%2Fportrait-2.png&w=384&q=75"
        srcset="/_next/image?url=%2Fassets%2Fportrait-2.png&w=256&q=75 1x, /_next/image?url=%2Fassets%2Fportrait-2.png&w=384&q=75 2x"
        style="color: transparent;"
        width="140"
      />
      <img
        alt="My profile picture"
        class="c3"
        data-nimg="1"
        decoding="async"
        height="140"
        loading="lazy"
        src="/_next/image?url=%2Fassets%2Fportrait-3.png&w=384&q=75"
        srcset="/_next/image?url=%2Fassets%2Fportrait-3.png&w=256&q=75 1x, /_next/image?url=%2Fassets%2Fportrait-3.png&w=384&q=75 2x"
        style="color: transparent;"
        width="140"
      />
      <img
        alt="My profile picture"
        class="c4"
        data-nimg="1"
        decoding="async"
        height="140"
        loading="lazy"
        src="/_next/image?url=%2Fassets%2Fportrait-4.png&w=384&q=75"
        srcset="/_next/image?url=%2Fassets%2Fportrait-4.png&w=256&q=75 1x, /_next/image?url=%2Fassets%2Fportrait-4.png&w=384&q=75 2x"
        style="color: transparent;"
        width="140"
      />
      <img
        alt="My profile picture"
        class="c5"
        data-nimg="1"
        decoding="async"
        height="140"
        loading="lazy"
        src="/_next/image?url=%2Fassets%2Fportrait-5.png&w=384&q=75"
        srcset="/_next/image?url=%2Fassets%2Fportrait-5.png&w=256&q=75 1x, /_next/image?url=%2Fassets%2Fportrait-5.png&w=384&q=75 2x"
        style="color: transparent;"
        width="140"
      />
      <img
        alt="My profile picture"
        class="c6"
        data-nimg="1"
        decoding="async"
        height="140"
        loading="lazy"
        src="/_next/image?url=%2Fassets%2Fportrait-6.png&w=384&q=75"
        srcset="/_next/image?url=%2Fassets%2Fportrait-6.png&w=256&q=75 1x, /_next/image?url=%2Fassets%2Fportrait-6.png&w=384&q=75 2x"
        style="color: transparent;"
        width="140"
      />
    </div>
  </div>
</body>
`;
